README.md
# SwiftPriorityQueue

[![Swift Versions](https://img.shields.io/badge/Swift-1%2C2%2C3%2C4%2C5-green.svg)](https://swift.org)
[![CocoaPods Version](https://img.shields.io/cocoapods/v/SwiftPriorityQueue.svg)](https://cocoapods.org/pods/SwiftPriorityQueue)
[![SPM Compatible](https://img.shields.io/badge/SPM-compatible-4BC51D.svg?style=flat)](https://swift.org/package-manager/)
[![CocoaPods Platforms](https://img.shields.io/cocoapods/p/SwiftPriorityQueue.svg)](https://cocoapods.org/pods/SwiftPriorityQueue)
[![Linux Compatible](https://img.shields.io/badge/Linux-compatible-4BC51D.svg?style=flat)](https://swift.org)
[![Twitter Contact](https://img.shields.io/badge/contact-@davekopec-blue.svg?style=flat)](https://twitter.com/davekopec)

SwiftPriorityQueue is a pure Swift (no Cocoa) implementation of a generic priority queue data structure, appropriate for use on all platforms (macOS, iOS, Linux, etc.) where Swift is supported. It features a straightforward interface and can be used with any type that implements `Comparable`. It utilizes comparisons between elements rather than separate numeric priorities to determine order.

Internally, SwiftPriorityQueue uses a classic binary heap, resulting in O(lg n) pushes and pops. It includes in-source documentation, an A* based example maze solving program (for macOS), and unit tests (*pull requests are welcome for additional unit tests in particular*).

## Features
* Easy-to-use method interface
* Small, self contained, pure Swift code base
* Classic binary heap implementation with O(lg n) pushes and pops
* Iterable through standard Swift for...in loops (implements `Sequence` and `IteratorProtocol`)
* In-source documentation
* A fun maze solving A* based example program

## Installation

Release 1.3.0 and above supports Swift 5. Use release 1.2.1 for Swift 4. Use release 1.1.2 for Swift 3 and Swift 2 support. Use release 1.0 for Swift 1.2 support.

### CocoaPods

Use the CocoaPod `SwiftPriorityQueue`.

### Swift Package Manager (SPM)

Add this repository as a dependency.

### Manual

Copy `SwiftPriorityQueue.swift` into your project.

## Documentation
There is a large amount of documentation in the source code using the standard Swift documentation technique (compatible with Jazzy).  Essentially though, SwiftPriorityQueue has the three critical methods you'd expect - `push()`, `pop()`, and `peek()`.

### Initialization
When you create a new `PriorityQueue` you can optionally specify whether the priority queue is ascending or descending. What does this mean? If the priority queue is ascending, its smallest values (as determined by their implementation of `Comparable` aka `<`) will be popped first, and if it's descending, its largest values will be popped first.
```
var pq: PriorityQueue<String> = PriorityQueue<String>(ascending: true)
```
You can also provide an array of starting values to be pushed sequentially immediately into the priority queue.
```
var pq: PriorityQueue<Int> = PriorityQueue<Int>(startingValues: [6, 2, 3, 235, 4, 500])
```
Or you can specify both.
```
var pq: PriorityQueue<Int> = PriorityQueue<Int>(ascending: false, startingValues: [6, 2, 3, 235, 4, 500])
```
Or you can specify neither. By default a `PriorityQueue` is descending and empty. As you've probably noticed, a PriorityQueue takes a generic type. This type must be `Comparable`, as its comparison will be used for determining priority.  This means that your custom types must implement `Comparable` and utilize the overridden `<` to determine priority.

### Methods
`PriorityQueue` has all of the standard methods you'd expect a priority queue data structure to have.
* `push(element: T)` - Puts an element into the priority queue. O(lg n)
* `push(element: T, maxCount: Int) -> T?` - Adds an element while limiting the size of the priority queue to `maxCount`. If more than `maxCount` elements are in the priority queue after the addition, the lowest priority element will be discarded and returned. Note this is inefficient because this is a binary heap, so only the highet priority item is efficient to retrieve. O(n)
* `pop() -> T?` - Returns and removes the element with the highest (or lowest if ascending) priority or `nil` if the priority queue is empty. O(lg n)
* `peek() -> T?` - Returns the element with the highest (or lowest if ascending) priority or `nil` if the priority queue is empty. O(1)
* `clear()` - Removes all elements from the priority queue.
* `remove(item: T)` - Removes the first found instance of *item* in the priority queue. Silently returns if not found. O(n)
* `removeAll(item: T)` - Removes all instances of *item* in the priority queue through repeated calls to `remove()`. Silently returns if not found.

### Properties
* `count: Int` - The number of elements in the priority queue.
* `isEmpty: Bool` - `true` if the priority queue has zero elements, and `false` otherwise.

### Standard Swift Protocols
`PriorityQueue` implements `IteratorProtocol`, `Sequence` and `Collection` so you can treat `PriorityQueue` like any other Swift sequence/collection. This means you can use Swift standard library fucntions on a `PriorityQueue` and iterate through a `PriorityQueue` like this:
```
for item in pq {  // pq is a PriorityQueue<String>
    print(item)
}
```
When you do this, every item from the `PriorityQueue` is popped in order. `PriorityQueue` also implements `CustomStringConvertible` and `CustomDebugStringConvertible`.
```
print(pq)
```
Note: `PriorityQueue` is *not* thread-safe (do not manipulate it from multiple threads at once).

### Limited Heap Size Example

Suppose you want to only keep the `maxCount` highest priority items in the priority queue.

For example, say you only want the priority queue to only ever have 4 elements:

```
var pq: PriorityQueue<Int> = PriorityQueue<Int>()
let maxCount = 4         

pq.push(4, maxCount: maxCount)
pq.push(5, maxCount: maxCount)
pq.push(0, maxCount: maxCount)
pq.push(3, maxCount: maxCount)  
pq.push(6, maxCount: maxCount)     
pq.push(1, maxCount: maxCount)     
```

In this case, after 4 elements were pushed, only the smallest elements were kept (because the order was `ascending`). So, the final priority queue has the elements 0, 1, 3, 4 in it. 

### Just for Fun - A* (`astar.swift`)
A* is a pathfinding algorithm that uses a priority queue. The sample program that comes with SwiftPriorityQueue is a maze solver that uses A*. You can find some in-source documentation if you want to reuse this algorithm inside `astar.swift`.

## Authorship & License
SwiftPriorityQueue is written by David Kopec (@davecom on GitHub) and released under the MIT License (see `LICENSE`). You can find my contact information on my GitHub profile page. I encourage you to submit pull requests and open issues here on GitHub. Thank you to all of the contributors over the years.


CHANGELOG.md
### 1.4.0
- Added a push() with a size limit, removing the lowest priority item (thanks @reedes)
- Added some asserts and tests (thanks @vale-cocoa)

### 1.3.1
- Improvements to pop() performance (thanks @peteraisher)
- Some basic performance tests (thanks @peteraisher)

### 1.3.0
- Swift 5 support

### 1.2.1
- Fixed a critical bug in remove() and added a test for it
- Rearranged the project to be testable on Linux
- Updated the format of Package.swift to be in-line with Swift 4

### 1.2.0
> Note: 1.2.0 is the first version of SwiftPriorityQueue to break compatibility with previous versions of Swift since release 1.0.1. From this point forward users of Swift 2 and Swift 3 should use release 1.1.2 of SwiftPriorityQueue.

- Swift 4 support
- Removed preprocessor macros and code to support Swift 2

### 1.1.2
- Initializer that takes custom order function added
- watchOS added to podspec

### 1.1.1
- Added remove(item: T) method to remove an item at an arbitrary location
- Added removeAll(item: T) method to remove multiple of the same item

### 1.1.0
- Swift 3 support

### 1.0.3
- Last Swift 2 only release
- Improved unit tests

### 1.0.2
- Better Swift 2 support

### 1.0.1
- Access control bug fix
- Documentation Additions

### 1.0
- Initial Stable Release
- Last Release to support Swift 1.2



Tests/LinuxMain.swift
import XCTest
@testable import SwiftPriorityQueueTests

XCTMain([
    testCase(SwiftPriorityQueueTests.allTests),
    testCase(SwiftPriorityQueuePerformanceTests.allTests),
])


Tests/SwiftPriorityQueueTests/SwiftPriorityQueueTests.swift
//
//  SwiftPriorityQueueTests.swift
//  SwiftPriorityQueue
//
//  Copyright (c) 2015-2023 David Kopec
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

import XCTest
import Foundation
@testable import SwiftPriorityQueue

// based on gist https://gist.github.com/rymcol/48a505c2a1c874daea52a296a2687f5f
// so we have something that sorta kinda looks like arc4random_uniform() on Linux
#if os(Linux)
    import SwiftGlibc
    
    public func arc4random_uniform(_ max: UInt32) -> Int32 {
        return (SwiftGlibc.rand() % Int32(max-1))
    }
#endif

class SwiftPriorityQueueTests: XCTestCase {

// Not sure who this test is relevant to. The Priority Queue only ensures orders of pops, other orders should not be
// depended upon
//    func testFastIteratingReturnsValuesInSameOrderOfIndexIteration() {
//        var pq = PriorityQueue<Int>(order: <, startingValues: [1, 2, 3, 4, 5])
//
//        var fastIterationValues = [Int]()
//        var iter = pq.makeIterator()
//        while let element = iter.next() {
//            fastIterationValues.append(element)
//        }
//
//        var indexIterationValues = [Int]()
//        for i in pq.startIndex..<pq.endIndex {
//            indexIterationValues.append(pq[i])
//        }
//
//        XCTAssertEqual(fastIterationValues, indexIterationValues)
//
//        // Let's also test with the other sort:
//        fastIterationValues.removeAll(keepingCapacity: true)
//        indexIterationValues.removeAll(keepingCapacity: true)
//        pq = PriorityQueue<Int>(order: >, startingValues: [5, 4, 3, 2, 1])
//
//        iter = pq.makeIterator()
//        while let element = iter.next() {
//            fastIterationValues.append(element)
//        }
//
//        var indexIterationValue = [Int]()
//        for i in pq.startIndex..<pq.endIndex {
//            indexIterationValue.append(pq[i])
//        }
//
//        XCTAssertEqual(fastIterationValues, indexIterationValue)
//    }
    
    func testIteratingViaIndexesValueSemantics() {
        var pq = PriorityQueue<Int>(order: <, startingValues: [1, 2, 3, 4, 5])
        var expectedHeap = pq.heap
        for i in pq.indices {
            let _ = pq[i]
        }
        XCTAssertEqual(pq.heap, expectedHeap)
        
        // Let's also test with the other sort:
        pq = PriorityQueue<Int>(order: >, startingValues: [5, 4, 3, 2, 1])
        expectedHeap = pq.heap
        for i in pq.indices {
            let _ = pq[i]
        }
        XCTAssertEqual(pq.heap, expectedHeap)
    }
    
    func testCustomOrder() {
        let priorities = [0: 5000, 1: 4000, 2: 3000, 3: 2000, 4: 1000, 5: 0]
        var pq: PriorityQueue<Int> = PriorityQueue<Int>(order: { priorities[$0]! > priorities[$1]! })
        for i in 0...5 {
            pq.push(i);
        }
        
        let expected: [Int] = [5, 4, 3, 2, 1, 0]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual)
    }
    
    func testBasic() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        for i in 0..<10 {
            pq.push(i);
        }
        
        let expected: [Int] = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual, "Basic 10 Integer Array Test Pass")
    }
    
    func testString() {
        var pq: PriorityQueue<String> = PriorityQueue<String>()
        var s = "a"
        while (s < "aaaaaa") {
            pq.push(s)
            s += "a"
        }
        
        let expected: [String] = ["aaaaa", "aaaa", "aaa", "aa", "a"]
        var actual: [String] = []
        for i in pq {
            actual.append(i)
        }
        
        XCTAssertEqual(expected, actual, "Basic 5 String Array Test Pass")
    }
    
    func testSetEquiv() {
        for _ in 0..<100 {
            var s = Set((0..<(arc4random_uniform(100))).map { _ in arc4random_uniform(1000000) })
            var q = PriorityQueue.init(startingValues: Array(s))
            XCTAssertEqual(s.count, q.count, "Incorrect count with elements: " + s.description)
            while let se = s.max() {
                XCTAssertEqual(se, q.pop(), "Incorrect max item with elements: " + s.description)
                s.remove(se)
            }
            XCTAssert(q.isEmpty, "Is not empty. Still contains: " + q.description)
        }
    }
    
    func testClear() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        for i in 0..<10 {
            pq.push(i);
        }
        pq.clear()
        XCTAssert(pq.isEmpty, "Is not empty. Still contains: " + pq.description)
    }
    
    func testPeek() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        pq.push(1)
        pq.push(5)
        pq.push(3)
        XCTAssertEqual(pq.peek(), 5, "Peek didn't return top element: " + pq.description)
    }
    
    func testRemove() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        for i in 0..<10 {
            pq.push(i);
        }
        
        pq.remove(4)
        pq.remove(7)
        
        let expected: [Int] = [9, 8, 6, 5, 3, 2, 1, 0]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual, "Trouble Removing 4 or 7")
    }
    
    func testRemoveAll() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        for i in 0..<10 {
            pq.push(i);
        }
        
        pq.push(4)
        pq.push(7)
        pq.push(7)
        
        pq.remove(4)
        pq.removeAll(7)
        
        let expected: [Int] = [9, 8, 6, 5, 4, 3, 2, 1, 0]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual, "Trouble Removing 4 or all 7s")
    }
    
    func testRemoveLastInHeap() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        pq.push(1)
        pq.push(2)
        
        pq.remove(1)
        
        let expected: [Int] = [2]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual)
    }

    func testPushWithLimitAscending() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>(ascending: false)
        let maxCount = 4
        XCTAssertNil(pq.push(4, maxCount: maxCount))
        XCTAssert(Set([4]) == Set(pq))
        XCTAssertNil(pq.push(5, maxCount: maxCount))
        XCTAssert(Set([4, 5]) == Set(pq))
        XCTAssertNil(pq.push(0, maxCount: maxCount))
        XCTAssert(Set([0, 4, 5]) == Set(pq))
        XCTAssertNil(pq.push(3, maxCount: maxCount))
        XCTAssert(Set([0, 3, 4, 5]) == Set(pq))
        XCTAssertEqual(pq.push(6, maxCount: maxCount), 6) // check first real discard
        XCTAssert(Set([0, 3, 4, 5]) == Set(pq))
        XCTAssertEqual(5, pq.push(1, maxCount: maxCount)) // check second discard
        print(pq)
        XCTAssert(Set([0, 1, 3, 4]) == Set(pq))
    }
    
    func testPushWithLimitDescending() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>(ascending: true)
        let maxCount = 4

        XCTAssertNil(pq.push(4, maxCount: maxCount))
        XCTAssert(Set([4]) == Set(pq))

        XCTAssertNil(pq.push(5, maxCount: maxCount))
        XCTAssert(Set([4, 5]) == Set(pq))

        XCTAssertNil(pq.push(2, maxCount: maxCount))
        XCTAssert(Set([2, 4, 5]) == Set(pq))

        XCTAssertNil(pq.push(3, maxCount: maxCount))
        XCTAssert(Set([2, 3, 4, 5]) == Set(pq))

        XCTAssertEqual(1, pq.push(1, maxCount: maxCount))
        XCTAssert(Set([2, 3, 4, 5]) == Set(pq))

        XCTAssertEqual(2, pq.push(6, maxCount: maxCount))
        XCTAssert(Set([3, 4, 5, 6]) == Set(pq))

        XCTAssertEqual(3, pq.push(6, maxCount: maxCount))
        XCTAssert(Set([4, 5, 6, 6]) == Set(pq))

    }
    
    static var allTests = [
        ("testCustomOrder", testCustomOrder),
        ("testBasic", testBasic),
        ("testString", testString),
        ("testSetEquiv", testSetEquiv),
        ("testClear", testClear),
        ("testPeek", testPeek),
        ("testRemove", testRemove),
        ("testRemoveAll", testRemoveAll),
        ("testRemoveLastInHeap", testRemoveLastInHeap),
        ("testPushWithLimitAscending", testPushWithLimitAscending),
        ("testPushWithLimitDescending", testPushWithLimitDescending),
        ]
}


Tests/SwiftPriorityQueueTests/SwiftPriorityQueuePerformanceTests.swift
//
//  SwiftPriorityQueuePerformanceTests.swift
//  SwiftPriorityQueue
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

import XCTest
@testable import SwiftPriorityQueue

class SwiftPriorityQueuePerformanceTests: XCTestCase {
    
    func testBuildPerformance() {
        let input: [Int] = Array((0 ..< 100000))
        measure {
            let _: PriorityQueue<Int> = PriorityQueue(ascending: true, startingValues: input)
        }
    }
    
    func testPopPerformance() {
        let original = PriorityQueue(ascending: true, startingValues: Array(0 ..< 100000))
        measure {
            var pq = original
            for _ in 0 ..< 100000 {
                let _ = pq.pop()
            }
        }
    }
    
    func testPushPerformance() {
        measure {
            for i in 0 ..< 100000 {
                var pq = PriorityQueue<Int>(ascending: true, startingValues: [])
                pq.push(i)
            }
        }
    }
    
    func testRemovePerformance() {
        let original = PriorityQueue(ascending: true, startingValues: Array(0 ..< 10000))
        measure {
            var pq = original
            for x in 0 ..< 100 {
                pq.remove(x * x)
            }
        }
    }
    
    static var allTests = [
        ("testBuildPerformance", testBuildPerformance),
        ("testPopPerformance", testPopPerformance),
        ("testPushPerformance", testPushPerformance),
        ("testRemovePerformance", testRemovePerformance)
    ]
}


SwiftPriorityQueue/astar.swift
//
//  astar.swift
//  SwiftPriorityQueue
//
//  Copyright (c) 2015-2019 David Kopec
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.

// This is an example of astar that uses SwiftPriorityQueue

class Node<T: Hashable>: Comparable, Hashable {
    let state: T
    let parent: Node?
    let cost: Float
    let heuristic: Float
    init(state: T, parent: Node?, cost: Float, heuristic: Float) {
        self.state = state
        self.parent = parent
        self.cost = cost
        self.heuristic = heuristic
    }
    
    func hash(into hasher: inout Hasher) {
        hasher.combine(state)
        hasher.combine(parent)
        hasher.combine(cost)
        hasher.combine(heuristic)
    }
}

func < <T>(lhs: Node<T>, rhs: Node<T>) -> Bool {
    return (lhs.cost + lhs.heuristic) < (rhs.cost + rhs.heuristic)
}

func == <T>(lhs: Node<T>, rhs: Node<T>) -> Bool {
    return lhs === rhs
}

/// Formulate a result path as an array from a goal node found in an astar search
///
/// - parameter startNode: The goal node found from an astar search.
/// - returns: An array containing the states to get from the start to the goal.
func backtrack<T>(_ goalNode: Node<T>) -> [T] {
    var sol: [T] = []
    var node = goalNode
    
    while (node.parent != nil) {
        sol.append(node.state)
        node = node.parent!
    }
    
    sol.append(node.state)
    
    return sol
}

/// Find the shortest path from a start state to a goal state.
///
/// - parameter initialState: The state that we are starting from.
/// - parameter goalTestFn: A function that determines whether a state is the goal state.
/// - parameter successorFn: A function that finds the next states from a state.
/// - parameter heuristicFn: A function that makes an underestimate of distance from a state to the goal.
/// - returns: A path from the start state to a goal state as an array.
func astar<T: Hashable>(_ initialState: T, goalTestFn: (T) -> Bool, successorFn: (T) -> [T], heuristicFn: (T) -> Float) -> [T]? {
    var frontier = PriorityQueue(ascending: true, startingValues: [Node(state: initialState, parent: nil, cost: 0, heuristic: heuristicFn(initialState))])
    var explored = Dictionary<T, Float>()
    explored[initialState] = 0
    var nodesSearched: Int = 0
    
    while let currentNode = frontier.pop() {
        nodesSearched += 1
          // we know if there are still items, we can pop one
        let currentState = currentNode.state
        
        if goalTestFn(currentState) {
            print("Searched \(nodesSearched) nodes.")
            return backtrack(currentNode)
        }
        
        for child in successorFn(currentState) {
            let newcost = currentNode.cost + 1  //1 assumes a grid, there should be a cost function
            if (explored[child] == nil) || (explored[child]! > newcost) {
                explored[child] = newcost
                frontier.push(Node(state: child, parent: currentNode, cost: newcost, heuristic: heuristicFn(child)))
            }
        }
    }
    
    return nil
}


SwiftPriorityQueueTests/SwiftPriorityQueueTests.swift
//
//  SwiftPriorityQueueTests.swift
//  SwiftPriorityQueueTests
//
//  Created by David Kopec on 3/27/15.
//  Copyright (c) 2015-2017 Oak Snow Consulting. All rights reserved.
//

import XCTest
import Foundation

class SwiftPriorityQueueTests: XCTestCase {
    
    func testCustomOrder() {
        let priorities = [0: 5000, 1: 4000, 2: 3000, 3: 2000, 4: 1000, 5: 0]
        var pq: PriorityQueue<Int> = PriorityQueue<Int>(order: { priorities[$0]! > priorities[$1]! })
        for i in 0...5 {
            pq.push(i);
        }
        
        let expected: [Int] = [5, 4, 3, 2, 1, 0]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual)
    }
    
    func testBasic() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        for i in 0..<10 {
            pq.push(i);
        }
        
        let expected: [Int] = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual, "Basic 10 Integer Array Test Pass")
    }
    
    func testString() {
        var pq: PriorityQueue<String> = PriorityQueue<String>()
        var s = "a"
        while (s < "aaaaaa") {
            pq.push(s)
            s += "a"
        }
        
        let expected: [String] = ["aaaaa", "aaaa", "aaa", "aa", "a"]
        var actual: [String] = []
        for i in pq {
            actual.append(i)
        }
        
        XCTAssertEqual(expected, actual, "Basic 5 String Array Test Pass")
    }
    
    func testSetEquiv() {
        for _ in 0..<100 {
            var s = Set((0..<(arc4random_uniform(100))).map { _ in arc4random_uniform(UInt32.max) })
            var q = PriorityQueue.init(startingValues: Array(s))
            XCTAssertEqual(s.count, q.count, "Incorrect count with elements: " + s.description)
            while let se = s.max() {
                XCTAssertEqual(se, q.pop(), "Incorrect max item with elements: " + s.description)
                s.remove(se)
            }
            XCTAssert(q.isEmpty, "Is not empty. Still contains: " + q.description)
        }
    }
    
    func testClear() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        for i in 0..<10 {
            pq.push(i);
        }
        pq.clear()
        XCTAssert(pq.isEmpty, "Is not empty. Still contains: " + pq.description)
    }
    
    func testPeek() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        pq.push(1)
        pq.push(5)
        pq.push(3)
        XCTAssertEqual(pq.peek(), 5, "Peek didn't return top element: " + pq.description)
    }
    
    func testRemove() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        for i in 0..<10 {
            pq.push(i);
        }
        
        pq.remove(4)
        pq.remove(7)
        
        let expected: [Int] = [9, 8, 6, 5, 3, 2, 1, 0]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual, "Trouble Removing 4 or 7")
    }
    
    func testRemoveAll() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        for i in 0..<10 {
            pq.push(i);
        }
        
        pq.push(4)
        pq.push(7)
        pq.push(7)
        
        pq.remove(4)
        pq.removeAll(7)
        
        let expected: [Int] = [9, 8, 6, 5, 4, 3, 2, 1, 0]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual, "Trouble Removing 4 or all 7s")
    }
    
    func testRemoveLastInHeap() {
        var pq: PriorityQueue<Int> = PriorityQueue<Int>()
        pq.push(1)
        pq.push(2)
        
        pq.remove(1)
        
        let expected: [Int] = [2]
        var actual: [Int] = []
        for j in pq {
            actual.append(j)
        }
        
        XCTAssertEqual(expected, actual)
    }
    
}
